Ethon & Typhoeus tracing support

al-kudryavtsev opened the pull request:
## Goal
Support distributed tracing when making requests through `Ethon` library (including `Typhoeus` requests which use `Ethon`)  
The related issue is https://github.com/DataDog/dd-trace-rb/issues/527

### Implementation details

Links to the libraries:
https://github.com/typhoeus/ethon 
https://github.com/typhoeus/typhoeus

The implementation was initially based on `rest_client` integration, but I had to change the approach because `Ethon` is a wrapper around `libcurl` which can execute many requests concurrently. I had to patch quite a few methods on `Easy` to support more use cases. There are different examples of the requests in the spec (made through Ethon's `Easy` & `Multi` and through Typhoeus's `Request` & `Hydra`).

This code has been in production for a few weeks, didn't see any issues. However we don't use concurrent requests (done through `Multi` or `Hydra`) so relying on unit tests for other use cases. `Ethon` also supports non-http requests, this PR does not address these use cases.

Here is a screenshot of distributed tracing achieved through new integration
![Screen Shot 2019-06-27 at 4 20 47 PM](https://user-images.githubusercontent.com/1422963/60626766-d89a1d00-9db9-11e9-8967-009b39c01803.jpg)

al-kudryavtsev commented:
Fixed ruby 1.9 support & rebased to get rid of failed mongo tests

delner commented:
@al-kudryavtsev The pull request is looking quite filled out! Is this ready for review?

Also, since I'm not familiar with Typhoeus/Ethon, could you explain the relationship between the two? What are the responsibilities of each library? Would a user ever want to activate Typhoeus tracing without Ethon or vice versa?

I saw Typhoeus added as a dependency, and Ethon getting some instrumentation, but I didn't see any instrumentation explicitly defined for Typhoeus, and it wasn't clear what this means for Typhoeus support.

al-kudryavtsev commented:
@delner Yes, the PR is ready for review, thanks!

Ethon is a `libcurl` wrapper which provides two main objects, Easy and Multi, which can be used to make requests (not only HTTP). Typhoeus is build on top of Ethon and is designed for HTTP requests. It adds extra features such as memoization / caching, and in general provides more high-level interface.

Initially I wanted to support Typhoeus because that's what we use for HTTP requests. The problem is that with extra features it is actually harder to implement proper & reliable instrumentation, especially with multiple parallel asynchronous requests. At the same time, Ethon's Easy object is a good candidate for tracking asynchronous requests because it is used when request starts and when response is ready.

What I did instead is instrumenting Ethon. This way HTTP requests coming through Typhoeus are fully supported and other use cases of Ethon are also supported (or support can be added relatively easy). I added Typhoeus as a dependency in order to unit test the instrumentation with Typhoeus primitives.

So, to answer your question, if the user wants to instrument HTTP requests made with Typhoeus, they will need to enable Ethon tracing. If they need to trace some other features of Typhoeus (like caching of responses), they will need a different instrumentation.

Also I'm not an author of any of these libraries, so it is possible I'm missing something and there is an easier way. There are hooks in both libraries, however it doesn't seem to be a reliable approach to tracing.

Please let me know what you think.

delner commented:
@al-kudryavtsev This is a super helpful description, thank you!

Sounds like it'll be good then to have this Ethon instrumentation either way, only major question to me outstanding is: "does this constitute 'Typhoeus' support?" I don't think I have this answer; I asked the folks on #527 what they think.

If it does, then we somehow have to make it obvious/intuitive to users who want Typhoeus traces to activate the Ethon instrumentation. This might require us to add something like `c.use :typhoeus` for that purpose, but we'll cross that bridge later.

If it doesn't, I will happily accept this PR as Ethon instrumentation on its own merits, but we won't be able to close #527, and as a community we'll need to find a better way to give the proper feature coverage to Typhoeus at a later date.

In any case, I'll give this a review a bit later!

che-burashco commented:
@delner Did you have a chance to look at the PR by any chance?

delner commented on lib/ddtrace/contrib/ethon/easy_patch.rb:
```diff
@@ -0,0 +1,146 @@
+          end
+
+          def datadog_before_request
+            @datadog_span = datadog_configuration[:tracer].trace(
```
What's the thought for using an instance variable here? It looks like maybe we need something to work for sync and async requests?

How does this work in a multithreaded application, or one that makes multiple async calls? My thought when I saw the use of an instance variable for this was: "can this conflict with other Ethon requests that are being made in parallel? This state would presumably be shared across all requests through the same instance of Ethon."

delner replied:
Okay, I think that's the key part; we don't leave unfinished spans (which will break traces), and we don't overwrite or accidentally re-use these spans for different operations.

delner commented on lib/ddtrace/contrib/ethon/easy_patch.rb:
```diff
@@ -0,0 +1,146 @@
+            datadog_tag_request
+
+            if datadog_configuration[:distributed_tracing]
+              Datadog::HTTPPropagator.inject!(@datadog_span.context, @datadog_original_headers)
```
Why store the original headers then inject distributed tracing here? Could you `inject!` into `self.headers` without using an instance variable?

delner replied:
I see; the basic thrust of my comment was to pursue simplicity, so if that suggestion wouldn't work or introduce more complexity, then I think this is fine.

delner commented on lib/ddtrace/contrib/ethon/easy_patch.rb:
```diff
@@ -0,0 +1,146 @@
+          def set_attributes(options)
+            return super unless tracer_enabled?
+
+            # Make sure headers= will get called
```
What's the reason for forcing `headers=` to be called?

delner commented on lib/ddtrace/contrib/ethon/multi_patch.rb:
```diff
@@ -0,0 +1,39 @@
+            handles = super(easy)
+            return handles if handles.nil? || !tracer_enabled?
+
+            easy.datadog_before_request
```
Not very familiar with the construction of the framework, but is `Multi` composed of several `Easy` instances? And when it executes, will it call `Easy#perform`? If so, would that call `datadog_before_request` making this redundant?

delner replied:
Per my response above, is there an equivalent `#perform` method for `Multi`? Might want to consider adding a parent span around the lifetime of this multi operation which bundles up the `Easy` requests. Just a thought.

al-kudryavtsev replied:
I added the parent span support. The logic is the following: it is created when the first `easy` is added to `multi` and it is finalized when `multi#perform` has finished. `Multi` can be reused - every time `easy` is added it creates a parent span if it doesn't exist yet.

delner commented on spec/ddtrace/contrib/ethon/ethon_patch_spec.rb:
```diff
@@ -0,0 +1,356 @@
+require 'stringio'
+require 'webrick'
+
+RSpec.describe Datadog::Contrib::Ethon do
```
There's a lot of good tests in here, but I think I'd like to see it broken out into smaller specs if possible. Generally speaking I like to see separate unit tests from the integration tests, then any other test subjects into their own specs. In this case something like:

 - `integration_spec.rb`: Integration tests for Ethon (mostly what you have here)
 - `typhoeus_integration_spec.rb`: Integration tests that use Typhoeus (what you have at the bottom)
 - `easy_spec.rb`: Unit tests for the Easy patch
 - `multi_spec.rb` Unit tests for the Multi patch

In the unit tests, I would try to mock/stub to avoid any external web calls (which tend to flake out in CI.) Using WEBrick in the integration tests is cool with me, but we should probably add `before { skip unless ENV['TEST_DATADOG_INTEGRATION'] }` at the top of them so they don't run in non-integration safe environments.

delner commented on spec/ddtrace/contrib/ethon/ethon_patch_spec.rb:
```diff
@@ -0,0 +1,356 @@
+    Datadog.registry[:ethon].reset_configuration!
+  end
+
+  describe 'instrumented request' do
```
Tests hit all the major areas, very nice :+1:

delner commented on lib/ddtrace/contrib/ethon/easy_patch.rb:
```diff
@@ -0,0 +1,146 @@
+          if Gem::Version.new(RUBY_VERSION) < Gem::Version.new('2.0.0')
+            base.class_eval do
+              # Among the patched methods, only this one is defined on Easy class, others are included
+              alias_method :set_attributes_without_datadog, :set_attributes
```
Sounds good! If you use `prepend`, make sure the version requirement for 2.0+ is reflected in the `Integration` definition. There are some examples of other integrations that do this.

al-kudryavtsev commented:
@delner I updated the PR - dropped ruby 1.9 support, removed `set_attributes` patch & improved tests

delner commented:
@al-kudryavtsev Cool, I'll review this again soon. Can you also rebase this against `0.26-dev`? That would be our merge target.

al-kudryavtsev commented:
@delner 👍 Rebased

delner commented on lib/ddtrace/contrib/ethon/easy_patch.rb:
```diff
@@ -0,0 +1,146 @@
+        module InstanceMethods
+          include InstanceMethodsCompatibility unless Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.0.0')
+
+          def http_request(url, action_name, options = {})
```
Okay this is pretty interesting, and exactly the kind of explanation I was looking for. (Thank you!)

I can see why you'd want to store this info before it gets turned into a format that's hard to read from; as long these objects are 1-1 with requests, then this seems okay. Just have to be careful in any scenario in which an `Easy` object is re-used or re-ran? Not sure if that happens, but some food for thought.

If `Multi` uses multiple `Easy` objects, would it make sense to add an additional parent span to the multi operation? Thinking about the case if users want to see the batch as an operation, in addition to its constituent parts.

al-kudryavtsev replied:
> I can see why you'd want to store this info before it gets turned into a format that's hard to read from; as long these objects are 1-1 with requests, then this seems okay. Just have to be careful in any scenario in which an Easy object is re-used or re-ran? Not sure if that happens, but some food for thought.

That's a good point. I added extra safety net by patching the [`reset`](https://github.com/typhoeus/ethon/blob/master/lib/ethon/easy.rb#L249) method which is used before re-using easy instance. This way I can be sure that HTTP method or headers are not mistakenly passed in the next request.

delner commented:
@al-kudryavtsev This is looking great; very happy with the thoroughness here, and the tests! Just left one more suggestion regarding `Multi`, but it's not an outright requirement, just a thought. Otherwise, I think we're in good shape to merge this.

al-kudryavtsev commented:
@delner Thanks! I made some changes to improve the correctness and added parent span support.

delner commented on lib/ddtrace/contrib/ethon/multi_patch.rb:
```diff
@@ -19,19 +19,48 @@ def add(easy)
             return handles if handles.nil? || !tracer_enabled?
 
-            easy.datadog_before_request
+            easy.datadog_before_request(parent_span: datadog_multi_span)
```
Is it possible there's a delay between `#add` and `#perform`? When the span is created, it's auto created with the start time. I'm thinking about the possibility where someone creates a `Multi` request with `Easy` children, but then delays a while before actually executing it; could that cause the timing to be recorded incorrectly?

al-kudryavtsev replied:
It is possible, more precise solution would probably require storing some state on `Multi` object. The complexity comes from the fact that extra `Easy` object can be added to `Multi` during the execution of `perform` (in easy complete callback, this is how Typhoeus implementation works). Current solution will work for Typhoeus since it always adding `easy` objects to `multi` right before executing it.
The same issue as you described can happen with `Easy` span because it is started when easy object is added to `multi` (which is not necessarily executing yet).
We can probably solve the issue by storing some flag on `multi` saying whether it is performing now or not.
Please let me know what you think. I'll be away from computer for one more week and will be able to work on this feature afterwards.

delner replied:
Might be worth it, as long as there isn't too much coupling between `Easy` and `Multi` instrumentation.

Is there any way to create the span at `perform` time for `Easy`? Even if that means storing other state on `Easy` initialization beforehand? Given these objects sound idempotent anyway, some state would be permissible.

al-kudryavtsev replied:
It turned out extra state is not needed. `Easy` handles are stored on `Multi` instances so I can go through them and start child spans. When adding `easy` to `multi`, I can check if there is a span on `multi` which is an indication of `multi`'s `perform` execution in progress.

delner commented:
@al-kudryavtsev I noticed the builds broke on Mongo: you'll probably want to rebase this PR onto the top of `0.27-dev`, which should fix the issue.

al-kudryavtsev commented:
@delner I added the change to make `multi` tracing more accurate, rebased it on top of `0.27-dev`. I also wrote a [test](https://gist.github.com/al-kudryavtsev/4f0de0d4b21d94722adaa4cc9144cef7) to manually test instrumentation.
Here is a screenshot of the `Hydra` instrumentation which starts new requests during underlying `multi`'s `perform` call: 
![Screen Shot 2019-08-19 at 5 36 33 PM](https://user-images.githubusercontent.com/1422963/63302252-a03b9900-c2aa-11e9-9abf-6995c4e3a371.png)

delner commented:
Wow, that screenshot looks awesome! I'll take another pass over the code, see if we can get this merged.

delner commented on lib/ddtrace/contrib/ethon/easy_patch.rb:
```diff
@@ -0,0 +1,136 @@
+
+            # Store headers to call this method again when span is ready
+            @datadog_original_headers = headers
+            super headers
```
Minor but can you just call `super` here?

delner commented on lib/ddtrace/contrib/ethon/easy_patch.rb:
```diff
@@ -0,0 +1,136 @@
+          def datadog_tag_request
+            span = @datadog_span
+            uri = URI.parse(url)
+            method = defined?(@datadog_method) ? @datadog_method.to_s : ''
```
Should this be `instance_variable_defined?`?

delner commented on lib/ddtrace/contrib/ethon/easy_patch.rb:
```diff
@@ -0,0 +1,136 @@
+            span = @datadog_span
+            uri = URI.parse(url)
+            method = defined?(@datadog_method) ? @datadog_method.to_s : ''
+            span.resource = "#{method} #{uri.path}".lstrip
```
We only want to set the HTTP method here, e.g. `GET`, but omit the path.

Sounds crazy, and I see the value of including the path. However, the resource is meant to be a `GROUP BY` like key, which in practice requires values that meaningfully group traces on some user-defined dimension. HTTP paths often contain unique input (such as numeric IDs, API tokens), which tend to flatten these traces into groups too small to be meaningful. Hence, as a default behavior, we've been only setting the method as the resource for HTTP client integrations.

If the omission of `path` as a default behavior is an issue for users, the concession we could offer is the option to customize their span and set their own resource using a callback. See HTTP as an example: https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/contrib/http/instrumentation.rb#L80. Implementing something like this is optional though, not necessary to merge this PR.

Also, it looks like the `method` can be `nil`; if you remove the path, it looks like the resource could then be an empty string, which will cause the trace to be malformed and dropped. We'll want to make sure resource is never blank, even if we have to give it some kind of default value instead.

al-kudryavtsev replied:
Interesting, I was wondering why the resource contains only method name in other instrumentations. I changed the default value of `method` to `N/A` & handled the case when it is `nil`.

delner commented on spec/ddtrace/contrib/ethon/easy_patch_spec.rb:
```diff
@@ -0,0 +1,194 @@
+  let(:configuration_options) { { tracer: tracer } }
+
+  before do
+    Datadog::Contrib::Ethon::Patcher.patch
```
I don't think this should be necessary: `c.use` will call `patch` automatically for the integration, and I think we want to verify that users don't need to manually patch.

delner commented on spec/ddtrace/contrib/ethon/multi_patch_spec.rb:
```diff
@@ -0,0 +1,134 @@
+  let(:configuration_options) { { tracer: tracer } }
+
+  before do
+    Datadog::Contrib::Ethon::Patcher.patch
```
Same as above.

al-kudryavtsev commented:
@delner Thanks for the review! I addressed the comments.
